### DISCUSSION TITLE
execvp error on file cardinal-opt (No such file or directory)

### DISCUSSION THREAD

**Discussion by zsh151** [neams-th-coe/cardinal#471] [Q&A] [answered]
Hello, everyone. Here I am again.

I try to follow Tutorial 8A to run the coupled OpenMC-THM-MOOSE calculation. Here is the new error.
```
en@ubuntu:~/cardinal/tutorials/gas_compact_multiphysics$ mpiexec -np 2 cardinal-opt -i common_input.i openmc_thm.i --n-threads=36
[proxy:0:0@ubuntu] HYDU_create_process (utils/launch/launch.c:74): execvp error on file cardinal-opt (No such file or directory)
[proxy:0:0@ubuntu] HYDU_create_process (utils/launch/launch.c:74): execvp error on file cardinal-opt (No such file or directory)
```

**Comment by aprilnovak** [comment, marked as answer]
Hi @zsh151, that error just means that your system could not find `cardinal-opt` ("No such file or directory"). You either need to have `cardinal-opt` on your `PATH` (by setting `export PATH=/home/en/cardinal:$PATH`), or use the actual path to the executable, like `mpiexec -np 2 ../../cardinal-opt -i common_input.i openmc_thm.i --n-threads=36`.

**Reply by zsh151** [reply]
Yeah, thanks. I later realized the problem.
Now the coupled OpenMC-THM-MOOSE calculation could run smoothly.But the coupled OpenMC-NekRS-MOOSE calculation cannot.
```
en@ubuntu:~/cardinal/tutorials/gas_compact_multiphysics$ mpiexec -np 500 cardinal-opt -i common_input.i openmc_nek.i
[proxy:0:0@ubuntu] HYDU_create_process (utils/launch/launch.c:22): pipe error (Too many open files)
[proxy:0:0@ubuntu] launch_procs (pm/pmiserv/pmip_cb.c:702): create process returned error
[proxy:0:0@ubuntu] HYD_pmcd_pmip_control_cmd_cb (pm/pmiserv/pmip_cb.c:885): launch_procs returned error
[proxy:0:0@ubuntu] HYDT_dmxu_poll_wait_for_event (tools/demux/demux_poll.c:77): callback returned error status
[proxy:0:0@ubuntu] main (pm/pmiserv/pmip.c:200): demux engine error waiting for event
[mpiexec@ubuntu] control_cb (pm/pmiserv/pmiserv_cb.c:208): assert (!closed) failed
[mpiexec@ubuntu] HYDT_dmxu_poll_wait_for_event (tools/demux/demux_poll.c:77): callback returned error status
[mpiexec@ubuntu] HYD_pmci_wait_for_completion (pm/pmiserv/pmiserv_pmci.c:196): error waiting for event
[mpiexec@ubuntu] main (ui/mpich/mpiexec.c:336): process manager error waiting for completion
```

**Reply by aprilnovak** [reply]
Does your system have 500 cores? The Nek case is quite large for this problem because the Reynolds number is high (so the mesh has many elements). You will probably not be able to run it on a typical personal computer, but there are many examples of HPC job scripts [on this page](https://cardinal.cels.anl.gov/hpc.html) which you can adapt for whatever cluster you have access to.

Alternatively, I recommend trying out the [pin bundle](https://cardinal.cels.anl.gov/tutorials/cht2.html) tutorial - this one you should be able to run on a personal computer.

**Reply by zsh151** [reply]
I see, thanks for your detailed explanation.

**Comment by zsh151** [comment]
Hi @aprilnovak,
May I ask if you have any other coupled OpenMC-NekRS-MOOSE calculation example that I can run on a personal computer?
I want to learn about the coupled OpenMC-NekRS-MOOSE calculation.

Thanks

**Reply by aprilnovak** [reply]
Hi @zsh151 - I added a bit more description to the [tutorials homepage](https://cardinal.cels.anl.gov/tutorials/index.html) that describes the interaction between NekRS and OpenMC - as long as you know how each one individually integrates with MOOSE, then you will have all the knowledge needed to couple them. Nevertheless, it is a good suggestion to have a small tutorial that does couple OpenMC and Nek at the same time.

I'm working on a new tutorial, and will let you know when it's completed. I'm traveling this week, so it probably won't be until next week.

**Reply by zsh151** [reply]
Thank you very much for all you've done. This will help me a lot
Hope you will enjoy your trip.

**Reply by aprilnovak** [reply]
Hi @zsh151 - I've just created a new tutorial that couples OpenMC, NekRS, and MOOSE heat conduction for a pincell. This example is small enough to run on most machines, and can be accessed here: https://cardinal.cels.anl.gov/tutorials/pincell_multiphysics.html

You will need to update Cardinal in order to pull in the newest files:

```
cd cardinal
git checkout devel
git pull origin devel
```

Thanks for your patience, it took me a bit longer than expected to find the time to put this together.
